segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameReader.java

```diff
@@ -104,13 +127,23 @@ public DataFrameRecord<T> getNext() throws DataCorruptionException, DurableDataL
                     T logItem = this.serializer.deserialize(this.dataFrameInputStream);
                     DataFrameRecord.RecordInfo recordInfo = this.dataFrameInputStream.endRecord();
                     long seqNo = logItem.getSequenceNumber();
+                    trackLastReadLogItem(seqNo, recordInfo);
                     if (seqNo <= this.lastReadSequenceNumber) {
-                        throw new DataCorruptionException(String.format("Invalid Operation Sequence Number. Expected: larger than %d, found: %d.",
-                                this.lastReadSequenceNumber, seqNo));
+                        // We have found a repeated item while reading the log. We need to discriminate between a harmless
+                        // duplicate entry or some other kind of data corruption.
+                        if (isHarmlessDuplicateEntry(seqNo, recordInfo)) {
```

Comment:
since this is done for every item read during recovery, would this affect recovery times in case of large recoveries?

Reply:
This is not done for each item: as you can see, this is inside the `if (seqNo <= this.lastReadSequenceNumber)` block, so this will be only executed in the exceptional case that we find a duplicate.